api/src/main/java/com/velocitypowered/api/proxy/server/QueryResponse.java

```diff
@@ -0,0 +1,306 @@
+            this.gameVersion = Preconditions.checkNotNull(gameVersion, "gameVersion");
+            return this;
+        }
+
+        @NonNull
+        public Builder map(@NonNull String map) {
+            this.map = Preconditions.checkNotNull(map, "map");
+            return this;
+        }
+
+        @NonNull
+        public Builder currentPlayers(int currentPlayers) {
+            Preconditions.checkArgument(currentPlayers >= 0, "currentPlayers cannot be negative");
+            this.currentPlayers = currentPlayers;
+            return this;
+        }
+
+        @NonNull
+        public Builder maxPlayers(int maxPlayers) {
+            Preconditions.checkArgument(maxPlayers >= 0, "maxPlayers cannot be negative");
+            this.maxPlayers = maxPlayers;
+            return this;
+        }
+
+        @NonNull
+        public Builder proxyHost(@NonNull String proxyHost) {
+            this.proxyHost = Preconditions.checkNotNull(proxyHost, "proxyHost");
+            return this;
+        }
+
+        @NonNull
+        public Builder proxyPort(int proxyPort) {
+            Preconditions.checkArgument(proxyPort >= 1 && proxyPort <= 65535, "proxyPort must be between 1-65535");
+            this.proxyPort = proxyPort;
+            return this;
+        }
+
+        @NonNull
+        public Builder players(@NonNull Collection<String> players) {
+            this.players.addAll(Preconditions.checkNotNull(players, "players"));
+            return this;
+        }
+
+        @NonNull
+        public Builder players(@NonNull String... players) {
+            this.players.addAll(Arrays.asList(Preconditions.checkNotNull(players, "players")));
+            return this;
+        }
+
+        @NonNull
+        public Builder clearPlayers() {
+            this.players.clear();
+            return this;
+        }
+
+        @NonNull
+        public Builder proxyVersion(@NonNull String proxyVersion) {
+            this.proxyVersion = Preconditions.checkNotNull(proxyVersion, "proxyVersion");
+            return this;
+        }
+
+        @NonNull
+        public Builder plugins(@NonNull Collection<PluginInformation> plugins) {
+            this.plugins.addAll(Preconditions.checkNotNull(plugins, "plugins"));
+            return this;
+        }
+
+        @NonNull
+        public Builder plugins(@NonNull PluginInformation... plugins) {
+            this.plugins.addAll(Arrays.asList(Preconditions.checkNotNull(plugins, "plugins")));
+            return this;
+        }
+
+        @NonNull
+        public Builder clearPlugins() {
+            this.plugins.clear();
+            return this;
+        }
+
+        /**
+         * Builds new {@link QueryResponse} with supplied data
+         * @return response
+         */
+        @NonNull
+        public QueryResponse build() {
+            return new QueryResponse(hostname, gameVersion, map, currentPlayers, maxPlayers, proxyHost, proxyPort, players, proxyVersion, plugins);
```

Comment:
To make this object truly immutable you also need to make the `plugins` and `players` list unmodifiable. The practice in Velocity is to use `ImmutableList.copyOf()` from Google Guava, which makes a full, compact copy of the collection.

I strongly recommend reading [_Effective Java_, 3rd edition](https://www.amazon.com/Effective-Java-3rd-Joshua-Bloch/dp/0134685997).

Reply:
More like I just forgot about them.